src/backend/distributed/deparser/citus_ruleutils.c

```diff
@@ -1353,3 +1354,139 @@ RoleSpecString(RoleSpec *spec, bool withQuoteIdentifier)
+
+			/* now inspect the refexpr that SubscriptingRef at hand were holding */
+			expr = refexpr;
+		}
+
+		if (expandedTargetEntries == NIL)
+		{
+			/* return original entry since it doesn't hold a SubscriptingRef node */
+			newTargetEntryList = lappend(newTargetEntryList, targetEntry);
+		}
+		else
+		{
+			/*
+			 * Need to concat expanded target list entries in reverse order
+			 * to preserve ordering of the original target entry list.
+			 */
+			newTargetEntryList = list_concat(newTargetEntryList,
+											 list_reverse(expandedTargetEntries));
+		}
+	}
+
+	return newTargetEntryList;
+}
+
+
+/*
+ * TargetListEntryExprFindSubscriptingRef searches given Expr --assuming that
+ * it is part of a target list entry-- to see if it directly (i.e.: itself)
+ * or indirectly (e.g.: behind some level of coercions) holds a SubscriptingRef
+ * node.
+ *
+ * Returns the original SubscriptingRef node on success or NULL otherwise.
+ *
+ * Note that it wouldn't add much value to use expression_tree_walker here
+ * since we are only interested in a subset of the fields of a few certain
+ * node types.
+ */
+static SubscriptingRef *
+TargetListEntryExprFindSubscriptingRef(Expr *expr)
+{
+	Node *node = (Node *) expr;
+	while (node)
+	{
+		if (IsA(node, FieldStore))
+		{
+			/*
+			 * ModifyPartialQuerySupported doesn't allow INSERT/UPDATE via
+			 * FieldStore. If we decide supporting such commands, then we
+			 * should take the first element of "newvals" list into account
+			 * here. This is because, to support such commands, we will need
+			 * to expand merged FieldStore into separate target entries too.
+			 *
+			 * For this reason, this block is not reachable atm and need to
+			 * uncomment the following if we decide supporting such commands.
+			 *
+			 * """
+			 *   FieldStore *fieldStore = (FieldStore *) node;
+			 *   node = (Node *) linitial(fieldStore->newvals);
+			 * """
+			 */
+			ereport(ERROR, (errmsg("unexpectedly got FieldStore object when "
+								   "generating shard query")));
+		}
+		else if (IsA(node, CoerceToDomain))
+		{
+			/*
+			 * Similarly, if FieldStore nodes were expected here, then it would
+			 * be possible to see implicit coercions on front of them too, but
+			 * is not the case atm. Need to uncomment the following too if we
+			 * decide supporting such commands.
+			 *
+			 * """
+			 *   CoerceToDomain *coerceToDomain = (CoerceToDomain *) node;
+			 *   if (coerceToDomain->coercionformat != COERCE_IMPLICIT_CAST)
+			 *   {
+			 *     // not an implicit coercion, cannot reach to a SubscriptingRef
+			 *     break;
+			 *   }
+			 *   node = (Node *) coerceToDomain->arg;
+			 * """
+			 */
+			ereport(ERROR, (errmsg("unexpectedly got CoerceToDomain object when "
+								   "generating shard query")));
+		}
+		else if (IsA(node, SubscriptingRef))
+		{
+			return (SubscriptingRef *) node;
+		}
+		else
+		{
+			break;
+		}
+	}
+
+	return NULL;
```

Comment:
Maybe get rid of this whole function since we cannot encounter with a `FieldStore`/`CoerceToDomain` node here atm, but I don't want to lose the findings noted here, let me think of a better way to document that.

Also, it's good to have those error checks as extra seat belts here.